### PR TITLE
UGENE-7455 Assert on Digest into fragments

### PR DESCRIPTION
**Pull request by DmitriiSukhomlinov**
https://ugene.dev/tracker/browse/UGENE-7455

### CODE REVIEW

**Review by smlnkcv — DISMISSED** [review]
No crash, no assert, error occurs in the report:
Error: | Conserved annotation Misc. Feature (2646..3236) is disrupted by the digestion. Try changing the restriction sites.

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp [hunk 1]
```diff
@@ -1380,6 +1384,69 @@ GUI_TEST_CLASS_DEFINITION(test_7451) {
...
+
+    // 3. Right click -> Cloning -> Digest into fragments...
+    class DigestScenario : public CustomScenario {
+    public:
+        void run(HI::GUITestOpStatus& os) override {
+            // 4. Select "AaaI" and click "Add---->"
+            // 5. Go to the "Conserved annotations" tab
+            QWidget* dialog = GTWidget::getActiveModalWidget(os);
+            QListWidget* availableEnzymeWidget = GTWidget::findExactWidget<QListWidget*>(os, "availableEnzymeWidget", dialog);
+            CHECK_SET_ERR(availableEnzymeWidget != nullptr, "Cannot find available enzyme list widget");
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: You do not need this check - findExactWidget makes it for you.

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp [hunk 2]
```diff
@@ -1380,6 +1384,69 @@ GUI_TEST_CLASS_DEFINITION(test_7451) {
     GTUtilsStartPage::checkRecentListUrl(os, "test_7451.fa", false);
 }
 
+GUI_TEST_CLASS_DEFINITION(test_7455) {
+    // 1. Open "_common_data/regression/7455/clipboard.gb"
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp:1388]: nit: numbering is inconvenient to support when you edit (add more steps) into the test. I would suggest never use it.

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp [hunk 3]
```diff
@@ -1380,6 +1384,69 @@ GUI_TEST_CLASS_DEFINITION(test_7451) {
...
+    GTUtilsTaskTreeView::waitTaskFinished(os);
+
+    // 3. Right click -> Cloning -> Digest into fragments...
+    class DigestScenario : public CustomScenario {
+    public:
+        void run(HI::GUITestOpStatus& os) override {
+            // 4. Select "AaaI" and click "Add---->"
+            // 5. Go to the "Conserved annotations" tab
+            QWidget* dialog = GTWidget::getActiveModalWidget(os);
+            QListWidget* availableEnzymeWidget = GTWidget::findExactWidget<QListWidget*>(os, "availableEnzymeWidget", dialog);
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: nit: type is typed twice in the same line

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp [hunk 4]
```diff
@@ -1380,6 +1384,69 @@ GUI_TEST_CLASS_DEFINITION(test_7451) {
...
+            GTWidget::click(os, GTWidget::findWidget(os, "addAnnBtn", dialog));
+
+
+
+            GTUtilsDialog::clickButtonBox(os, dialog, QDialogButtonBox::Ok);
+        }
+    };
+
+    GTLogTracer lt;
+    GTUtilsDialog::waitForDialog(os, new DigestSequenceDialogFiller(os, new DigestScenario));
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: Use () for constructors (the common style in UGENE)

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp [hunk 5]
```diff
@@ -1380,6 +1384,69 @@ GUI_TEST_CLASS_DEFINITION(test_7451) {
...
+                    auto selectAnnotationsList = GTWidget::findWidgetByType<QListWidget*>(os, dialog, "Cant find the \"Select annotations\" list");
+                    auto items = GTListWidget::getItems(os, selectAnnotationsList);
+                    CHECK_SET_ERR(items.size() == 1, "Unexpected number of annotations");
+
+                    GTListWidget::click(os, selectAnnotationsList, items.first());
+                    GTUtilsDialog::clickButtonBox(os, dialog, QDialogButtonBox::Ok);
+                }
+            };
+
+            GTUtilsDialog::waitForDialog(os, new Filler(os, "select_annotations_dlalog", new SelectAnnotationScenario));
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: Use () for constructors (the common style in UGENE)

File: src/plugins/enzymes/src/DigestSequenceDialog.cpp [hunk 6]
```diff
@@ -157,9 +157,11 @@ void DigestSequenceDialog::accept() {
     for (int row = 0; row < itemCount; ++row) {
         const QString &annEncoded = conservedAnnsWidget->item(row)->text();
         QStringList annData = annEncoded.split(" ");
-        assert(annData.size() == 2);
-        QString aName = annData.at(0);
-        QString locationStr(annData.at(1));
+        assert(annData.size() >= 2);
```
Comment (yalgaer) [inline src/plugins/enzymes/src/DigestSequenceDialog.cpp]: We do not use asserts anymore, we use SAFE_POINTS

File: src/plugins/enzymes/src/DigestSequenceDialog.cpp [hunk 7]
```diff
@@ -157,9 +157,11 @@ void DigestSequenceDialog::accept() {
...
+
+        QString regionString = annData.last();
```
Comment (yalgaer) [inline src/plugins/enzymes/src/DigestSequenceDialog.cpp]: You are making the same error here that caused the bug: you are parsing visual representation of data and this process leads to errors. 
First between name and location there may be more spaces (this is a visual formatting only), next the location itself may contain spaces (Genbank format allows it).

What I would suggest as a correct fix: 
1) Store annotation data you need (name & location or the whole annotation data) as QVariant in the QListWidgetItem
2) Take and use it directly with no parsing at all.

**Review by yalgaer — CHANGES_REQUESTED** [review]

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp (hunk 1 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: OK

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp (hunk 2 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp:1388]: The ordered list looks better for me

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp (hunk 3 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: OK

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp (hunk 4 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: OK

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp (hunk 5 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: OK

File: src/plugins/enzymes/src/DigestSequenceDialog.cpp (hunk 7 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/enzymes/src/DigestSequenceDialog.cpp]: Fixed

File: src/plugins/enzymes/src/DigestSequenceDialog.cpp (hunk 6 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/enzymes/src/DigestSequenceDialog.cpp]: Removed at all

**Review by yalgaer — COMMENTED** [review]
Looks good now. Will approve after GUI tests are finished.

**Review by yalgaer — APPROVED** [review]

**Pull request merged**